modules/core/react/lib/InputProvider.tsx

```diff
@@ -169,48 +158,63 @@ export default class InputProvider extends React.Component<InputProviderProps, I
   // Need to remember how it component was mounted so we ensure listener removal
   provideIntent = this.props.provideIntent;
 
+  getContainer(container: HTMLElement | React.RefObject<HTMLElement> | undefined): HTMLElement {
+    if (container === undefined) {
+      return document.body;
+    } else if ('current' in container) {
+      if (container.current === null) {
+        console.warn('Your ref object can not be null, therefore, falling back to document.body');
+        return document.body;
+      } else {
+        return container.current;
+      }
+    }
+    return container;
+  }
+
+  private container = this.getContainer(this.props.container);
+
   componentDidMount() {
+    if (this.container.hasAttribute('data-whatinput')) {
+      return;
+    }
+
     // For IE11 and under, we'll need to polyfill element.closest
     elementClosestPolyfill(window);
-
-    if (
-      this.ref.current &&
-      this.ref.current.parentElement &&
-      this.ref.current.parentElement.closest(`.${containerClass}`)
-    ) {
-      this.setState({nested: true});
+    if (this.container.parentElement && this.container.parentElement.closest('[data-whatinput]')) {
```

Comment:
If you change this to `this.container.closest('[data-whatinput]'))`, you can remove the if statement above.

This will set `nested` to `true` if the container is already "decorated", but that's okay. The name `nested` doesn't really cover all cases anymore anyway. The intent of this variable was to track whether there was another InputProvider doing its thing. Since it's now possible for this to be true regardless of whether they are nested or not, we should probably rename this variable.